### DISCUSSION TITLE
Add mask

### DISCUSSION THREAD

**Discussion by cotegeo** [maplibre/maplibre-gl-js#3237] [Q&A]
Good morning,

I'd like to put a mask layer on my app. I saw on a forum it exists a property 'fill-region' with values 'inside' and "outside". But this doesn't work. The console returns "unknown property". Is there a property that enable to make mask ?

Thanks

Sylvain BECQUAERT

**Comment by HarelM** [comment]
The slack channel is better for these kind of questions, but I'm not sure I fully understand the requirements.
Maybe a mock up of what you want to do can help here.
Also I would advise to consult the spec docs here:
https://maplibre.org/maplibre-style-spec/

**Comment by bameyrick** [comment]
This is a feature I would also be very much interested in, and would like to help to implement - but would likely need support in understanding how MapLibre works. 

### Use case

I'm not sure if my use case is the same as what @cotegeo has in mind but we've been rendering Cloud Optimised GeoTiffs to a raster layer using [@geomatico/maplibre-cog-protocol](https://www.npmjs.com/package/@geomatico/maplibre-cog-protocol). We have functionality where we only want to render data within user selected country boundaries, but because the COG has data for the whole world we want a mask layer above the COG layer to hide everything but the selected countries.

| Unmasked | Masked |
| ---------- | -------- |
| <img  alt="Screenshot 2025-01-16 at 15 31 48" src="https://github.com/user-attachments/assets/2547b409-e48b-4726-86d0-44e1d181f064" /> | <img  alt="Screenshot 2025-01-16 at 15 32 50" src="https://github.com/user-attachments/assets/d74e34d3-1e04-419c-b059-cb8a751273fd" /> |

<hr/>

### Masking via a custom layer

I recently managed to create a proof of concept masking layer that implements [CustomLayerInterface](https://maplibre.org/maplibre-gl-js/docs/API/interfaces/CustomLayerInterface/). It's quite a rudimentary solution that only works with mercator projection, but it would be great to have this natively supported as I think it could run much more efficiently and maybe also support globe projection.

The custom layer:

1. Accepts an array of `Feature<Polygon>`.
2. Converts each polygon to vertices (using [earcut](https://www.npmjs.com/package/earcut)).
3. Uses stencil buffering to draw the polygon vertices and then extract them from a quad that covers the entire screen.

There is [example code in this discussion](https://github.com/maplibre/maplibre-gl-js/discussions/5353) where I was struggling with the stencil buffer but then found a fudge to fix the rendering issues that I won't bother duplicating here, but when I was bumbling around `maplibre-gl-js`'s source code could sort of understand that there are classes and methods of things like [stencil mode](https://github.com/maplibre/maplibre-gl-js/blob/main/src/gl/stencil_mode.ts), [subdivision](https://github.com/maplibre/maplibre-gl-js/blob/main/src/render/subdivision.ts), etc that seem like the sort of things needed to create a mask layer (just in a much more optimised way).

<hr/>

### Proposal

As @cotegeo mentioned, I think it would be good to add a `fill-region` paint option to the existing fill layer where you can specify `outside` or `inside` where inside is the default and existing functionality of a fill layer, and outside creates a mask by painting the inverse.

This would be good as theoretically the masking could work with a [VectorTileSource](https://maplibre.org/maplibre-gl-js/docs/API/classes/VectorTileSource/) or [GeoJSON source](https://maplibre.org/maplibre-style-spec/sources/#geojson) and even use [filter expressions](https://maplibre.org/maplibre-style-spec/expressions/) to pick what is not masked out.


#### Potential issues

The issue with this approach is that the mask is just a painted colour so if you had raster tiles below as your base map they would be hidden too.

#### Alternative approach

Maybe a better solution would be to add masking to [raster layers](https://maplibre.org/maplibre-style-spec/layers/#raster) to only paint the pixels of the raster within the provided polygons, but I have no idea if that is even possible, and if being able to provide raster layer with a [vector source](https://maplibre.org/maplibre-style-spec/sources/#vector) to mask with alongside the [raster source](https://maplibre.org/maplibre-style-spec/sources/#raster) would be messy.

<hr/>

Although I am keen to help add functionality like this I wouldn't even know where to start or what the best approach to take would be, but I am happy to try figure it out with some guidance 😊

**Comment by HarelM** [comment]
Is this a dynamic change you need to do? Otherwise you could:
1. Change the data
2. Change the plugin to support clipping

**Reply by bameyrick** [reply]
Yes it would need to be dynamic as users can select any combination of admin boundaries which we then subtract from the mask.

By  
>  Change the plugin to support clipping

do you mean the cog protocol plugin used for loading the tiles?

**Reply by HarelM** [reply]
Yes.
I'm not sure how such masking can be dynamic without affecting all layers etc.
In any case, feel free to open an issue with a spec change proposal in the maplibre-style-spec repo.
Please provide as many information as possible and also sdd it to the monthly meeting agenda.

**Reply by bameyrick** [reply]
Cool, thank you @HarelM. I have been experimenting with achieving masking in the protocol too, so I will continue with that and I can use what I find to feed into the spec change proposal 👍
